src/fft/mod.rs

```diff
@@ -0,0 +1,410 @@
+//!     complex64::new(0.0, 0.0),
+//!     complex64::new(0.0, 0.0),
+//! ];
+//! assert_eq!(fft2_result.as_slice::<complex64>(), &expected[..]);
+//!
+//! let mut ifft2_result = ifft2(&fft2_result, None, None);
+//! ifft2_result.eval();
+//! assert_eq!(ifft2_result.dtype(), Dtype::Complex64);
+//!
+//! let expected = &[
+//!    complex64::new(1.0, 0.0),
+//!    complex64::new(1.0, 0.0),
+//!    complex64::new(1.0, 0.0),
+//!    complex64::new(1.0, 0.0),
+//! ];
+//! assert_eq!(ifft2_result.as_slice::<complex64>(), &expected[..]);
+//!
+//! let mut rfft2_result = rfft2(&array, None, None);
+//! rfft2_result.eval();
+//! assert_eq!(rfft2_result.dtype(), Dtype::Complex64);
+//!
+//! let expected = &[
+//!     complex64::new(4.0, 0.0),
+//!     complex64::new(0.0, 0.0),
+//!     complex64::new(0.0, 0.0),
+//!     complex64::new(0.0, 0.0),
+//! ];
+//! assert_eq!(rfft2_result.as_slice::<complex64>(), &expected[..]);
+//!
+//! let mut irfft2_result = irfft2(&rfft2_result, None, None);
+//! irfft2_result.eval();
+//! assert_eq!(irfft2_result.dtype(), Dtype::Float32);
+//! assert_eq!(irfft2_result.as_slice::<f32>(), &src[..]);
+//!
+//! // The original array is not modified
+//! let data: &[f32] = array.as_slice();
+//! assert_eq!(data, &[1.0, 1.0, 1.0, 1.0]);
+//! ```
+//!
+//! ## `N` dimensions
+//!
+//! ```rust
+//! use mlx::{Dtype, Array, StreamOrDevice, complex64, fft::*};
+//!
+//! let array = Array::ones::<f32>(&[2, 2, 2]);
+//! let mut fftn_result = fftn(&array, None, None);
+//! fftn_result.eval();
+//! assert_eq!(fftn_result.dtype(), Dtype::Complex64);
+//!
+//! let mut expected = [complex64::new(0.0, 0.0); 8];
+//! expected[0] = complex64::new(8.0, 0.0);
+//! assert_eq!(fftn_result.as_slice::<complex64>(), &expected[..]);
+//!
+//! let mut ifftn_result = ifftn(&fftn_result, None, None);
+//! ifftn_result.eval();
+//! assert_eq!(ifftn_result.dtype(), Dtype::Complex64);
+//!
+//! let expected = [complex64::new(1.0, 0.0); 8];
+//! assert_eq!(ifftn_result.as_slice::<complex64>(), &expected[..]);
+//!
+//! let mut rfftn_result = rfftn(&array, None, None);
+//! rfftn_result.eval();
+//! assert_eq!(rfftn_result.dtype(), Dtype::Complex64);
+//!
+//! let mut expected = [complex64::new(0.0, 0.0); 8];
+//! expected[0] = complex64::new(8.0, 0.0);
+//! assert_eq!(rfftn_result.as_slice::<complex64>(), &expected[..]);
+//!
+//! let mut irfftn_result = irfftn(&rfftn_result, None, None);
+//! irfftn_result.eval();
+//! assert_eq!(irfftn_result.dtype(), Dtype::Float32);
+//!
+//! let expected = [1.0; 8];
+//! assert_eq!(irfftn_result.as_slice::<f32>(), &expected[..]);
+//!
+//! // The original array is not modified
+//! let data: &[f32] = array.as_slice();
+//! assert_eq!(data, &[1.0; 8]);
+//! ```
+
+mod fftn;
+mod rfftn;
+
+use smallvec::SmallVec;
+
+use crate::{
+    error::FftError,
+    utils::{all_unique, resolve_index, resolve_index_unchecked},
+    Array,
+};
+
+pub use self::{fftn::*, rfftn::*};
+
+#[inline]
+fn resolve_size_and_axis_unchecked(
```

Comment:
minor: I'd love to move these to an `fft/utils.rs` just to make this module easy to digest for those looking.